modules/hivemq-edge-module-opcua/src/main/java/com/hivemq/edge/adapters/opcua/OpcUaClientConnection.java

```diff
@@ -0,0 +1,350 @@
+        this.config = config;
+        this.tagStreamingService = tagStreamingService;
+        this.dataPointFactory = dataPointFactory;
+        this.eventService = eventService;
+        this.adapterId = adapterId;
+        this.protocolAdapterState = protocolAdapterState;
+        this.protocolAdapterMetricsService = protocolAdapterMetricsService;
+    }
+
+    private static byte @NotNull [] convertPayload(
+            final @NotNull DataValue dataValue,
+            final @NotNull EncodingContext serializationContext) {
+        //null value, empty buffer
+        if (dataValue.getValue().getValue() == null) {
+            return Constants.EMPTY_BYTES;
+        }
+
+        final ByteBuffer byteBuffer = OpcUaToJsonConverter.convertPayload(serializationContext, dataValue);
+        final ByteBuffer rewind = byteBuffer.asReadOnlyBuffer().rewind();
+        final byte[] array = new byte[rewind.remaining()];
+        rewind.get(array);
+        return array;
+    }
+
+    public @NotNull CompletableFuture<Void> start() {
+        ioLock.lock();
+        try {
+            final var result = ParsedConfig.fromConfig(config);
+            final ParsedConfig parsedConfig;
+
+            //this gets a lot nice with Java 21 and pattern matching.
+            if (result instanceof final Failure<ParsedConfig, String> failure) {
+                log.error("Failed to parse configuration for OPC UA client: {}", failure.failure());
+                return CompletableFuture.failedFuture(new IllegalArgumentException(failure.failure()));
+            } else if (result instanceof final Success<ParsedConfig, String> success) {
+                parsedConfig = success.result();
+            } else {
+                throw new IllegalStateException("Unexpected result type: " + result.getClass().getName());
+            }
+
+            final var configurator = new OpcUaClientConfigurator(adapterId, parsedConfig);
+
+            final OpcUaClient newOcpUaClient;
+            try {
+                newOcpUaClient = OpcUaClient.create(uri,
+                        endpoints -> endpoints.stream().findFirst(),
+                        transportConfigBuilder -> {},
+                        configurator);
+            } catch (final UaException e) {
+                log.error("Unable to create OpcUaClient", e);
+                return CompletableFuture.failedFuture(e);
+            }
+
+            return newOcpUaClient.connectAsync().thenCompose(same_as_newOcpUaClient -> {
+                log.info("Client created and connected successfully");
+                return createSubscription(newOcpUaClient);
+            }).thenApply(ignored -> (Void) null).whenComplete((ignored, throwable) -> {
+                if (throwable != null) {
+                    log.error("Failed to create subscription", throwable);
+                    protocolAdapterState.setConnectionStatus(ERROR);
+                    try {
+                        newOcpUaClient.disconnect();
```

Comment:
2 questions:

1. This lambda is not protected by the `iolock`. Please confirm this is expected behavior.
2. Is `disconnect()` call in all branches? Is it possible that it is not disconnected in edge cases?